state/pruner.go

```diff
@@ -210,6 +221,58 @@ func (p *Pruner) SetABCIResRetainHeight(height int64) error {
 	return nil
 }
 
+func (p *Pruner) SetTxIndexerRetainHeight(height int64) error {
+	// Ensure that all requests to set retain heights via the application are
+	// serialized.
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+	if height <= 0 {
+		return ErrInvalidHeightValue
+	}
+
+	currentRetainHeight, err := p.txIndexer.GetRetainHeight()
+	if err != nil {
+		if !errors.Is(err, ErrKeyNotFound) {
+			return err
+		}
+		return p.txIndexer.SetRetainHeight(height)
+	}
+	if currentRetainHeight > height {
+		return ErrPrunerCannotLowerRetainHeight
+	}
+	if err := p.txIndexer.SetRetainHeight(height); err != nil {
+		return err
+	}
+	// TODO call metrics
```

Comment:
Is this something we want to do in this PR?

Reply:
It will be done in a follow up PR.